lib/semantics/resolve-names.cc

```diff
@@ -3568,6 +3571,78 @@ bool DeclarationVisitor::Pre(const parser::StructureConstructor &x) {
+    const parser::ObjectName &pointeeName{std::get<1>(bp.t)};
+    auto *pointer{FindSymbol(pointerName)};
+    if (!pointer) {
+      pointer = &MakeSymbol(pointerName, ObjectEntityDetails{});
+    } else if (pointer->has<UseDetails>()) {
+      Say(pointerName,
+          "'%s' cannot be a Cray pointer as it is use-associated"_err_en_US);
+    } else if (!ConvertToObjectEntity(*pointer) || IsParameter(*pointer)) {
+      SayWithDecl(pointerName, *pointer, "'%s' is not a variable"_err_en_US);
+    } else if (pointer->Rank() > 0) {
+      SayWithDecl(pointerName, *pointer,
+          "Cray pointer '%s' must be a scalar"_err_en_US);
+    } else if (pointer->test(Symbol::Flag::CrayPointee)) {
+      Say(pointerName,
+          "'%s' cannot be a Cray pointer as it is already a Cray pointee"_err_en_US);
+    }
+    pointer->set(Symbol::Flag::CrayPointer);
+    const DeclTypeSpec &pointerType{MakeNumericType(TypeCategory::Integer,
+        context().defaultKinds().subscriptIntegerKind())};
+    const auto *type{pointer->GetType()};
+    if (!type) {
+      pointer->SetType(pointerType);
+    } else if (*type != pointerType) {
+      Say(pointerName.source, "Cray pointer '%s' must have type %s"_err_en_US,
+          pointerName.source, pointerType.AsFortran());
+    }
+    if (ResolveName(pointeeName)) {
+      Symbol &pointee{*pointeeName.symbol};
+      if (pointee.has<UseDetails>()) {
+        Say(pointeeName,
+            "'%s' cannot be a Cray pointee as it is use-associated"_err_en_US);
+      } else if (!ConvertToObjectEntity(pointee) || IsParameter(pointee)) {
+        Say(pointeeName, "'%s' is not a variable"_err_en_US);
+      } else if (pointee.test(Symbol::Flag::CrayPointer)) {
+        Say(pointeeName,
+            "'%s' cannot be a Cray pointee as it is already a Cray pointer"_err_en_US);
+      } else if (pointee.test(Symbol::Flag::CrayPointee)) {
+        Say(pointeeName,
+            "'%s' was already declared as a Cray pointee"_err_en_US);
+      } else {
+        pointee.set(Symbol::Flag::CrayPointee);
+      }
+      if (const auto *pointeeType{pointee.GetType()}) {
+        if (const auto *derived{pointeeType->AsDerived()}) {
+          if (!derived->typeSymbol().get<DerivedTypeDetails>().sequence()) {
+            Say(pointeeName,
+                "Type of Cray pointee '%s' is a non-sequence derived type"_err_en_US);
+          }
+        }
+      }
+      // process the pointee array-spec, if present
+      BeginArraySpec();
+      Walk(std::get<std::optional<parser::ArraySpec>>(bp.t));
+      const auto &spec{arraySpec()};
+      if (spec.empty()) {
+        // No array spec
+      } else if (pointee.Rank() > 0) {
+        SayWithDecl(pointeeName, pointee,
+            "Array spec was already declared for '%s'"_err_en_US);
+      } else if (!spec.IsExplicitShape() && !spec.IsAssumedSize()) {
+        Say(pointeeName,
+            "'%s' must have must have explicit shape or assumed size"_err_en_US);
```

Comment:
This check should probably also be done if the pointee shape is declare outside of the BasedPointeeStmt else the following will be accepted:
```
  function foo(x)
    real :: x(:)
    pointer(i, x)
  end function
```